xarray/tests/test_namedarray.py

```diff
@@ -39,27 +46,53 @@ def test_as_compatible_data_with_0d_object() -> None:
     np.array_equal(as_compatible_data(data), data)
 
 
-def test_as_compatible_data_with_explicitly_indexed(random_inputs) -> None:
+def test_as_compatible_data_with_explicitly_indexed(
+    random_inputs: np.ndarray[Any, Any]
+) -> None:
     # TODO: Make xr.core.indexing.ExplicitlyIndexed pass is_duck_array and remove this test.
-    class CustomArray(xr.core.indexing.NDArrayMixin):
-        def __init__(self, array):
+    class CustomArrayBase(xr.core.indexing.NDArrayMixin):
+        def __init__(self, array: T_DuckArray) -> None:
             self.array = array
 
-    class CustomArrayIndexable(CustomArray, xr.core.indexing.ExplicitlyIndexed):
+        @property
+        def dtype(self) -> np.dtype[np.generic]:
+            return self.array.dtype
+
+        @property
+        def shape(self) -> tuple[int, ...]:
+            return self.array.shape
+
+        @property
+        def real(self) -> Self:
+            raise NotImplementedError
+
+        @property
+        def imag(self) -> Self:
+            raise NotImplementedError
+
+        def astype(self, dtype: np.typing.DTypeLike) -> Self:
+            raise NotImplementedError
+
+    class CustomArray(CustomArrayBase):
+        def __array__(self) -> np.ndarray[Any, np.dtype[np.generic]]:
+            return np.array(self.array)
+
+    class CustomArrayIndexable(CustomArrayBase, xr.core.indexing.ExplicitlyIndexed):
```

Comment:
ExplicitlyIndexed has a non-standard `__array__` method. This made inheriting from it difficult.